Vhar 6679 hae bonukset sanktionakymaan

K1ll3rF0x opened the pull request:
WIP

jpluomajoki-solita committed:
siisti requireja

jpluomajoki-solita committed:
WIP: Hae bonukset mukaan sanktionäkymään

K1ll3rF0x committed:
Merge remote-tracking branch 'origin/develop' into VHAR-6679_hae-bonukset-sanktionakymaan

K1ll3rF0x committed:
Korjaa sanktiot.sql syntaksivirheet

RelNotes:
Related:

K1ll3rF0x committed:
Korjaa UNION ALL vastaavien sarakkeiden määrä

RelNotes:
Related:

K1ll3rF0x committed:
Näytä viiva tyhjän sanktion tyypin tai tapahtumapaikka/kuvaus kohdalla

RelNotes:
Related:

K1ll3rF0x committed:
YKsinkerstaista urakan bonusten hakua hakemalla bonukset erillisessä
kyselyssä
* Bonusten ja sanktioiden kyselyiden tulokset yhdistetään rajapinnan
toteutuksessa CLJ-koodissa.

RelNotes:
Related:

K1ll3rF0x committed:
Korjaa sanktiot-listauksessa laji :lupausbonus -> :lupaus-bonus

RelNotes:
Related:

K1ll3rF0x committed:
Säädä sanktiot-listauksen sarakkeiden leveyksiä, jotta isommat summat
mahtuvat paremmin näkyviin.

RelNotes:
Related:

K1ll3rF0x committed:
Nimeä Sanktiot-välilehti -> 'Sanktiot ja bonukset', sekä muuta näkymän
otsikko.

RelNotes:
Related:

K1ll3rF0x committed:
Nimeä hae-urakan-sanktiot rajapinta uudelleen, mahdollista rajapinnalla
erikseen sanktioiden tai bonusten haku. Palauta oletuksena sekä
sanktiot, että bonukset.

RelNotes:
Related:

K1ll3rF0x committed:
Nimeä symboli 'haetut-sanktiot' -> 'haetut-sanktiot-ja-bonukset'

RelNotes:
Related:

K1ll3rF0x committed:
Säädä indeksikorjausten sääntöjä urakan bonusten kyselyssä, lisää
TODO-kommentteja

RelNotes:
Related:

jpluomajoki-solita commented on src/clj/harja/kyselyt/sanktiot.sql:
```diff
@@ -130,6 +131,70 @@ WHERE
+       TRUE                   AS suorasanktio,
+       ek.toimenpideinstanssi AS toimenpideinstanssi,
+       CASE
+           -- FIXME Meillä voi olla lajeina 'lupausbonus' erilliskustannuksista tai 'lupaus-bonus' urakka_paatos
```
Onko tässä jotain korjattavaa vielä? Mielestäni tämä menee aivan oikein, erilliskustannuksissa voi olla vain lupausbonus, jota ei varmaan käytännössä edes koskaan tehdä erilliskustannuksena?

K1ll3rF0x replied:
Ongelma on siinä, että sanktiotyypeissä ei ole "lupaus-bonus" tai "lupaus-sanktio" lajeja. 

Kirjoitusasu "lupaus-bonus" on eri kuin kaikkialla muualla. Nyt tulee responsessa sekaisin erilliskustannuksista "lupausbonus" lajia ja toisaalta urakka_paatos-taulusta "lupaus-bonus" tyyppiä.

Kirjoitusasun tulisi olla yhtenevä "lupausbonus", jotta käyttöliittymässä ja muualla ei tarvi kahteen kertaan määritellä
:lupaus-bonus "Lupausbonus"
:lupausbonus "Lupausbonus"

ja jotta ne eivät menisi ylipäänsä sekaisin koodissa

K1ll3rF0x replied:
Varmaan myös halutaan näyttää vanhat urakoiden kirjaukset, jolloin reponseen voi tulla erilliskustannuksista "lupausbonus" tyypiisiä bonuksia.

Ellei kokonaan siirretä erilliskustannukset-taulusta vanhoja kirjauksia jonnekin muualle.

jpluomajoki-solita commented on src/clj/harja/kyselyt/sanktiot.sql:
```diff
@@ -130,6 +131,70 @@ WHERE
+       TRUE                                            AS suorasanktio,
+       NULL                                            AS toimenpideinstanssi, -- TODO Tarkista
+       CASE
+           -- FIXME Meillä voi olla lajeina 'lupausbonus' erilliskustannuksista tai 'lupaus-bonus' urakka_paatos taulusta
```
Sama kuin ylempi kommentti, eikös tämä ole nyt aivan oikein?

jpluomajoki-solita commented on src/clj/harja/kyselyt/sanktiot.sql:
```diff
@@ -130,6 +131,70 @@ WHERE
+           -- FIXME Meillä voi olla lajeina 'lupausbonus' erilliskustannuksista tai 'lupaus-bonus' urakka_paatos taulusta
+           WHEN p.tyyppi::TEXT IN ('lupaus-bonus')
+               THEN (SELECT korotus
+               -- FIXME speksin mukaan indeksikorotus pitää laskea MHU 2018 & 2020 lupausbonukselle.
```
2018 on varmaan typo? Indeksikorotus lasketaan 19-20 urakoille, 21 alkaen enää ei, kiitos `sanktion-indeksikorotus` - funktion.

jpluomajoki-solita replied:
Itse asiassa ilmeisesti yhdessä 2018 alkaneessa urakassa on lupauksia, joten tähän pitänee vielä korjata se, että siinä tapauskessa ei käytetä ollenkaan indeksiä. Varmistettava vielä ensin singletonilta, että oliko näin

K1ll3rF0x replied:
Jep, typotin sen vuosiluvun fixme-kommenttiin.
Ei tuossa minusta ollut edes korjattavaa, koska käytin mielestäni oikein tuota tietokantafunktiota.
Mutta, merkitsin muutoksen kommentilla, jotta se käytäisiin vielä läpi.

K1ll3rF0x committed:
Merge branch 'develop' into VHAR-6679_hae-bonukset-sanktionakymaan

jpluomajoki-solita commented on src/clj/harja/kyselyt/sanktiot.sql:
```diff
@@ -91,21 +91,22 @@ SELECT
+  t.toimenpidekoodi                   AS tyyppi_toimenpidekoodi,
+  t.koodi                             AS tyyppi_koodi
 
+-- FIXME Pitääkö hakea kaveriksi vielä 'lupaus-sanktio' urakka-paatos taulusta?
```
Mielestäni hyvä, nämä voisi hakea tässä kyselyssä.

jpluomajoki-solita commented on src/clj/harja/kyselyt/sanktiot.sql:
```diff
@@ -130,6 +131,70 @@ WHERE
+       TRUE                   AS suorasanktio,
+       ek.toimenpideinstanssi AS toimenpideinstanssi,
+       CASE
+           -- FIXME Meillä voi olla lajeina 'lupausbonus' erilliskustannuksista tai 'lupaus-bonus' urakka_paatos
```
Onko tässä jotain korjattavaa? Mielestäni menee aivan oikein nyt

jpluomajoki-solita commented on src/clj/harja/kyselyt/sanktiot.sql:
```diff
@@ -130,6 +131,70 @@ WHERE
+  FROM urakka_paatos p
+     JOIN urakka u ON u.id = p."urakka-id"
+ WHERE p."urakka-id" = :urakka
+   -- FIXME 'lupaus-sanktio' taitaa olla vielä erillinen sanktiotyyppi. Pitää varmaan hakea se ylempänä: 'hae-urakan-sanktiot'
```
Mielestäni validi fixme, lupaus-sanktio pois tästä ja ylempään kyselyyn

jpluomajoki-solita reviewed (commented):
Muutokset näyttää hyvältä, ja fixme-ehdotukset ovat mielestäni valideja.

Onkohan erilliskustannus- ja sanktiotaulujen lupausbonus ja lupaussanktio enää valideja? Voisi tarkistaa taustan, miksi ne on ylipäätään lisätty, ja voisiko ne kenties poistaa kokonaan, luultavasti niitä ei pitäisi pystyä tekemään erilliskustannuksista/sanktioista ollenkaan.

K1ll3rF0x committed:
Merge branch 'develop' into VHAR-6679_hae-bonukset-sanktionakymaan

jpluomajoki-solita committed:
Merge branch 'develop' into VHAR-6679_hae-bonukset-sanktionakymaan

jpluomajoki-solita committed:
Hae lupaussanktiot sanktioiden mukana, lisää niihin sanktion
indeksikorjaus

jpluomajoki-solita committed:
Yhtenäistä lupausbonusten ja -sanktioiden nimeäminen

 - muutettu välikatselmuksesta tehtävät lupaus-bonus/sanktio
   lupaubonus/sanktioksi, ilman viivaa.

jpluomajoki-solita committed:
Korjaa sanktiolajin näkyminen lukitussa laatupoikkeamassa

jpluomajoki-solita committed:
Merge branch 'develop' into VHAR-6679_hae-bonukset-sanktionakymaan

jpluomajoki-solita committed:
bump migraatio

jpluomajoki-solita committed:
Merge branch 'develop' into VHAR-6679_hae-bonukset-sanktionakymaan

jpluomajoki-solita committed:
Merge branch 'develop' into VHAR-6679_hae-bonukset-sanktionakymaan

K1ll3rF0x committed:
Merge branch 'develop' into VHAR-6679_hae-bonukset-sanktionakymaan

jlansineva committed:
Merge branch 'develop' into VHAR-6679_hae-bonukset-sanktionakymaan

rckopteri commented on src/clj/harja/kyselyt/sanktiot.sql:
```diff
@@ -128,7 +127,119 @@ WHERE
         OR
              lp.yllapitokohde IS NOT NULL AND
-             (SELECT poistettu FROM yllapitokohde WHERE id = lp.yllapitokohde) IS NOT TRUE);
+             (SELECT poistettu FROM yllapitokohde WHERE id = lp.yllapitokohde) IS NOT TRUE)
```
Saisiko tämän yksinkertaistettua riville 114 ->
LEFT JOIN yllapitokohde ypk ON lp.yllapitokohde = ypk.id AND ypk.poistettu IS NOT TRUE

Ja vielä parempi, jos poistettu kentälle antaisi default arvon ja pakottaisi arvot olemaan joko true tai false,
Ja sitten voisi tehdä vertailu ypk.poistettu = false. (tämä on suorituskyvyltään parempi)

jpluomajoki-solita replied:
Poistettu-check ei toimi joinissa, koska silloin rivi näkyy, vaikka sen ypk olisi poistettu. Ilmeisesti siinä tapauksessa ei haluta näyttää sanktiota ollenkaan.

is not true -> is false voidaan tehdä, sillä on jo default false.

rckopteri commented on src/clj/harja/palvelin/palvelut/laadunseuranta.clj:
```diff
@@ -123,9 +139,9 @@
-                                                         :loppu (konv/sql-timestamp loppu)}))]
+                   (concat
+                     urakan-sanktiot
+                     urakan-bonukset))]
```
oisko tarpeen järjestää tulokset tässä vaiheessa jotenkin?

jpluomajoki-solita replied:
Järjestäminen tehdään jo frontilla, niin mielestäni turha alkaa säätämään tässä vaiheessa.

K1ll3rF0x replied:
Mietitään tulosten järjestelyä tarkemmin siinä yhteydessä kun toteutetaan sanktioiden ja bonusten listaukseen filtteröinti.

rckopteri commented on test/clj/harja/palvelin/palvelut/laadunseuranta_test.clj:
```diff
@@ -558,10 +558,14 @@
+                                                                   ;; vertailua varten
+                                                                   :hae-sanktiot? true
+                                                                   :hae-bonukset? false})
         sanktio (first vastaus)
```
Ei haittaisi, vaikka tähän lisäisi tarkistuksen, että bonuksia ei tosiaan resultsetissä ole

jpluomajoki-solita replied:
Lisätty tarkistus

rckopteri commented on test/clj/harja/palvelin/palvelut/laadunseuranta_test.clj:
```diff
@@ -466,9 +466,9 @@
 (deftest hae-urakan-sanktiot
```
Tästä on vähän hankala löytää niin kysyn.
Oliko tässä tarkistettu, että yhtään bonusta löytyisi kannasta? 
Tuntuisi, että joku epäonnistuva ja happycase ois aina kiva olla mukana.

jpluomajoki-solita replied:
Tässä testissä ei ole, kyseisellä urakalla ei ole bonuksia. Lisätty sanktioiden tarkistus toiseen testiin

rckopteri reviewed (approved):
Jätin pari kommenttia. Jos on kiire mergetä, niin ei haittaa vaikka menee näinkin.

jpluomajoki-solita committed:
Parantele laadunseurannan testejä

jpluomajoki-solita committed:
is not true -> is false

jpluomajoki-solita committed:
Nosta migraation versio

jpluomajoki-solita committed:
Merge branch 'develop' into VHAR-6679_hae-bonukset-sanktionakymaan
